### PR TITLE
txnsync: Add linked list based short term lru cache

### PR DESCRIPTION
**Pull request by tsachiherman**
## Summary

Existing short term cache in the transaction sync was not supporting promoting updated cache entries.
In this PR, I have replaced the circulating buffer with a dual linked list to allow promoting cached entries.

## Test Plan

Extend existing unit test

### CODE REVIEW

File: txnsync/transactionCache.go [hunk 1]
```diff
@@ -169,31 +198,100 @@ func (lru *transactionCache) acknowledge(seqs []uint64) {
 
 // add a given transaction ID to the short term cache.
 func (st *shortTermTransactionCache) add(txid transactions.Txid) {
-	if st.transactionsMap[txid] {
+	entry, exists := st.transactionsMap[txid]
+	if exists {
+		// promote
+		if entry.detach() {
```
Comment (brianolson) [inline txnsync/transactionCache.go:204]: if `entry == st.first` we could just `st.first=entry.next` and not detach+insert it?

**Review by brianolson — APPROVED** [review]
LGTM. One possible rearrange

File: txnsync/transactionCache.go [hunk 2]
```diff
@@ -53,12 +53,21 @@ type ackPendingTxids struct {
 	timestamp time.Duration
 }
 
+// shortTermCacheEntry is used as the data container of a single linked list item
```
Comment (algorandskiy) [inline txnsync/transactionCache.go:56]: single linked list -> double linked list since it has next/prev

File: txnsync/transactionCache.go [hunk 3]
```diff
@@ -68,13 +77,33 @@ type longTermTransactionCache struct {
...
+	if ce.next == ce.prev {
+		return false
+	}
+	ce.prev.next = ce.next
+	ce.next.prev = ce.prev
+	return true
+}
+
+// addToList add the element to the tail of the list who's head is firstListEntry.
+func (ce *shortTermCacheEntry) addToList(firstListEntry *shortTermCacheEntry) {
```
Comment (algorandskiy) [inline txnsync/transactionCache.go:93]: firstListEntry -> head, as the most common name, so it would eliminate half of the comment :)

File: txnsync/transactionCache.go [hunk 4]
```diff
@@ -53,12 +53,21 @@ type ackPendingTxids struct {
...
+}
+
 //msgp:ignore shortTermTransactionCache
 type shortTermTransactionCache struct {
-	size            int
-	transactionsIDs []transactions.Txid
-	transactionsMap map[transactions.Txid]bool
-	oldest          int
+	size            int                                        // the maximum number of elements in the short term cache
+	first           *shortTermCacheEntry                       // pointer to first element
```
Comment (algorandskiy) [inline txnsync/transactionCache.go:68]: first -> head?
Also a comment that first/head is the oldest entry

File: txnsync/transactionCache.go [hunk 5]
```diff
@@ -169,31 +198,100 @@ func (lru *transactionCache) acknowledge(seqs []uint64) {
...
+		return
+	}
+	used := st.first
+	free := st.free
+	free.prev.next = used
+	used.prev.next = free
+	lastFree := free.prev
+	free.prev = used.prev
+	used.prev = lastFree
+	st.first = nil
```
Comment (algorandskiy) [inline txnsync/transactionCache.go:294]: could you please add a test that calls `reset` in 2 used and 3 free nodes and ensures `st.free` has all 5 nodes reachable in both directions?

File: txnsync/transactionCache.go (hunk 1 above)
Comment (algorandskiy) [reply txnsync/transactionCache.go:204]: sounds right but not sure if the code will become clearer.

**Review by algorandskiy — COMMENTED** [review]
Looks good, couple comments

File: txnsync/transactionCache.go (hunk 5 above)
Comment (tsachiherman) [reply txnsync/transactionCache.go:294]: sure.
